parley/src/layout/data.rs

```diff
@@ -470,4 +481,43 @@ impl<B: Brush> LayoutData<B> {
             }
         }
     }
+
+    fn calculate_content_widths(&mut self) {
+        let mut max_width = 0.0;
+        for item in &self.items {
+            match item.kind {
+                LayoutItemKind::TextRun => {
+                    let run = &self.runs[item.index];
+                    let mut min_width = 0.0;
+                    let mut trailing_whitespace = 0.0;
+                    for cluster in &self.clusters[run.cluster_range.clone()] {
+                        let boundary = cluster.info.boundary();
+                        if matches!(boundary, Boundary::Line | Boundary::Mandatory) {
+                            self.min_content_width =
+                                self.min_content_width.max(min_width - trailing_whitespace);
+                            min_width = 0.0;
+                            if boundary == Boundary::Mandatory {
```

Comment:
I wonder if `greedy.rs` can be rewritten to use `Boundary::Mandatory` rather than checking for newline white space.

Reply:
Maybe @dfrg can shine some light on this, but from my quick look at Swash, I assume that `Boundary::Mandatory` corresponds the mandatory break in the Unicode standard [here (see also Table 1)](https://www.unicode.org/reports/tr14/#LD7), which does seem to be the correct choice for line breaking as well.

Reply:
Swash follows the Unicode LBA and attaches `Boundary::Mandatory` to the cluster after the newline sequence which means we never encounter that state when the source text ends with a newline. This is arguably a swash bug but I “fixed” it in parley by just checking for the newline white space flag and that actually simplified a lot of code.